src/common/checksum.c

```diff
@@ -20,218 +20,509 @@
+	int rc = daos_csummer_prep_csum_buf(obj, rec_len, nr,
+		recxs, pcsum_bufs);
+
+	if (rc)
+		return rc;
+
+	daos_csum_buf_t		*csums = *pcsum_bufs;
+	uint8_t			*buf;
+	size_t			 bytes_for_csum;
+	size_t			 csum_nr;
+	uint64_t		 bytes;
+	uint32_t		 chunk_size = (uint32_t)obj->dcs_chunk_size;
+	uint32_t		 i;
+	uint32_t		 j;
+	struct daos_sgl_idx	 idx = {0};
+
+	for (i = 0; i < nr; i++) { /** for each extent/checksum buf */
+		csum_nr = daos_recx_calc_chunks(recxs[i], rec_len, chunk_size);
+		bytes = recxs[i].rx_nr * rec_len;
+
+		for (j = 0; j < csum_nr; j++) {
+			buf = dcb_idx2csum(&csums[i], j);
+			daos_csummer_set_buffer(obj, buf, csums->cs_len);
+			daos_csummer_reset(obj);
+
+			bytes_for_csum = MIN(chunk_size, bytes);
+			rc = daos_sgl_processor(sgl, &idx, bytes_for_csum,
+						checksum_sgl_cb, obj);
+			if (rc)
+				return rc;
+
+			daos_csummer_finish(obj);
+			bytes -= bytes_for_csum;
 		}
 	}
-failed:
-	return rc;
+
+	return 0;
+}
+
+void
+daos_csummer_destroy_csum_buf(struct daos_csummer *obj,
+			      daos_csum_buf_t **pcsum_buf)
+{
+	if (!(daos_csummer_initialized(obj) && *pcsum_buf))
+		return;
+	D_FREE((*pcsum_buf)->cs_csum);
+	D_FREE((*pcsum_buf));
 }
 
+/**
+ * daos_csum_buf_t functions
+ */
+void
+dcb_set(daos_csum_buf_t *csum_buf, void *buf,
+	uint32_t csum_buf_size,
+	uint16_t csum_size,
+	uint32_t csum_count,
+	uint32_t chunksize)
+{
+	csum_buf->cs_csum = buf;
+	csum_buf->cs_len = csum_size;
+	csum_buf->cs_buf_len = csum_buf_size;
+	csum_buf->cs_nr = csum_count;
+	csum_buf->cs_chunksize = chunksize;
+}
+
+void
+dcb_set_null(daos_csum_buf_t *csum_buf)
+{
+	dcb_set(csum_buf, NULL, 0, 0, 0, 0);
+}
+
+bool
+dcb_is_valid(const daos_csum_buf_t *csum)
+{
+	return csum != NULL &&
+	       csum->cs_len > 0 &&
+	       csum->cs_buf_len > 0 &&
+	       csum->cs_csum != NULL &&
+	       csum->cs_chunksize > 0 &&
+	       csum->cs_nr > 0;
+}
+
+uint32_t
+dcb_off2idx(daos_csum_buf_t *csum_buf, uint32_t offset_bytes)
+{
+	if (csum_buf->cs_chunksize == 0)
+		return 0;
+	return offset_bytes / csum_buf->cs_chunksize;
+}
+
+uint8_t *
+dcb_idx2csum(daos_csum_buf_t *csum_buf, uint32_t idx)
+{
+	uint32_t offset = csum_buf->cs_len * idx;
+
+	/** Make sure the buffer is big enough for the indexed checksum */
+	if (csum_buf->cs_buf_len < offset + csum_buf->cs_len)
+		return NULL;
+	return csum_buf->cs_csum + offset;
+}
+
+uint8_t *
+dcb_off2csum(daos_csum_buf_t *csum_buf, uint32_t offset)
+{
+	return dcb_idx2csum(csum_buf,
+			    dcb_off2idx(csum_buf, offset));
+}
+
+/** Other Functions */
+
+uint32_t
+daos_recx_calc_chunks(daos_recx_t extent, uint32_t record_size,
+		      uint32_t chunk_size)
+{
+	return (uint32_t) csum_chunk_count(chunk_size,
+					   extent.rx_idx,
+					   extent.rx_idx + extent.rx_nr - 1,
+					   record_size);
+}
+
+uint64_t
+csum_chunk_count(uint32_t chunk_size, uint64_t lo_idx, uint64_t hi_idx,
+		 uint64_t rec_size)
+{
+	if (chunk_size == 0)
+		return 0;
+	uint64_t lo = lo_idx * rec_size;
```

Comment:
Liang asked me to ensure C89...so put variable declarations at the top of each block rather than in the middle.

Reply:
Fixed